File: src/main/scala/com/lx/algos/nlp/nagao/TFNeighbor.scala
package com.lx.algos.nlp.nagao

import scala.collection.mutable.{HashMap, Map}

/**
  *
  * @project scalaML
  * @author lx on 4:52 PM 19/12/2017
  */

class TFNeighbor private[nagao]() {
  private val leftNeighbor:  Map[Char, Int] = new HashMap[Char, Int]().withDefaultValue (0)
  private val rightNeighbor: Map[Char, Int] = new HashMap[Char, Int]().withDefaultValue (0)
  private var tf = 0

  //add word to leftNeighbor
  def addToLeftNeighbor(word: Char): Unit = { //leftNeighbor.put(word, 1 + leftNeighbor.getOrDefault(word, 0));
    leftNeighbor(word) += 1
  }

  //add word to rightNeighbor
  def addToRightNeighbor(word: Char): Unit = { //rightNeighbor.put(word, 1 + rightNeighbor.getOrDefault(word, 0));
    rightNeighbor(word) += 1
  }

  //increment tf
  def incrementTF(): Unit = {
    tf += 1
  }

  def getLeftNeighborNumber: Int = leftNeighbor.size

  def getRightNeighborNumber: Int = rightNeighbor.size

  def getLeftNeighborEntropy: Double = {
    var entropy = 0.0
    var sum = 0.0
    for (number <- leftNeighbor.values) {
      entropy += number * Math.log(number)
      sum += number
    }
    if (sum == 0)  0.0
    else Math.log(sum) - entropy / sum
  }

  def getRightNeighborEntropy: Double = {
    var entropy = 0.0
    var sum = 0.0
    for (number <- rightNeighbor.values) {
      entropy += number * Math.log(number)
      sum += number
    }
    if (sum == 0) 0.0
    else Math.log(sum) - entropy / sum
  }

  def getTF: Int = tf
}

File: src/main/java/com/lx/nlp/nagao/TFNeighbor.java
package com.lx.nlp.nagao;

import java.util.HashMap;
import java.util.Map;

/**
 * @author lx on 10:56 AM 19/12/2017
 * @project scalaML
 */


public class TFNeighbor {

    private int tf;
    private Map<Character, Integer> leftNeighbor;
    private Map<Character, Integer> rightNeighbor;

    TFNeighbor(){
        leftNeighbor = new HashMap<Character, Integer>();
        rightNeighbor = new HashMap<Character, Integer>();
    }
    //add word to leftNeighbor
    public void addToLeftNeighbor(char word){
        //leftNeighbor.put(word, 1 + leftNeighbor.getOrDefault(word, 0));
        Integer number = leftNeighbor.get(word);
        leftNeighbor.put(word, number == null? 1: 1+number);
    }
    //add word to rightNeighbor
    public void addToRightNeighbor(char word){
        //rightNeighbor.put(word, 1 + rightNeighbor.getOrDefault(word, 0));
        Integer number = rightNeighbor.get(word);
        rightNeighbor.put(word, number == null? 1: 1+number);
    }
    //increment tf
    public void incrementTF(){
        tf++;
    }
    public int getLeftNeighborNumber(){
        return leftNeighbor.size();
    }
    public int getRightNeighborNumber(){
        return rightNeighbor.size();
    }
    public double getLeftNeighborEntropy(){
        double entropy = 0;
        int sum = 0;
        for(int number : leftNeighbor.values()){
            entropy += number*Math.log(number);
            sum += number;
        }
        if(sum == 0)  return 0;
        return Math.log(sum) - entropy/sum;
    }
    public double getRightNeighborEntropy(){
        double entropy = 0;
        int sum = 0;
        for(int number : rightNeighbor.values()){
            entropy += number*Math.log(number);
            sum += number;
        }
        if(sum == 0)  return 0;
        return Math.log(sum) - entropy/sum;
    }
    public int getTF(){
        return tf;
    }
}